File: dijkstra_4dir.m
function initial_path = dijkstra_4dir(Mmap, start_loc, goal_loc)
% This function just reports an initial path for calculating initial cable
% state


queue = [start_loc];

temp_map = Mmap;

father_index = zeros(size(Mmap, 1), size(Mmap, 2), 2);
father_index(queue(1, 1), queue(1, 2), :) = [-1, -1];
cost = inf(size(Mmap, 1), size(Mmap, 2));
cost(queue(1, 1), queue(1, 2)) = 0;


while 1
    x = queue(1, 1);
    y = queue(1, 2);
    queue(1, :) = [];
    sons = [x, y] + [-1, 0; 0, -1; 0, 1; 1, 0];

    if x == goal_loc(1) && y == goal_loc(2)
        initial_path = [];
        cur = [x, y];
        while 1
            if cur(1) == start_loc(1) && cur(2) == start_loc(2)
                break;
            end
            father = father_index(cur(1), cur(2), :);
            initial_path = [cur(1), cur(2); initial_path];
            cur = father(1, 1, :);
        end
        break;
    end
    
    for i = 1:size(sons, 1)
        son = sons(i, :);
        
        if Mmap(son(1), son(2)) == 254
            continue;
        end
            
        if cost(son(1), son(2)) ~= inf
            % We are using BFS. 
            % If a node has been explored, then we cannot have a shortcut
            continue;
        end
        
        father_index(son(1), son(2), :) = [x, y];
        cost(son(1), son(2)) = cost(x, y) + 1;
        
        queue = [queue; son];
    end
    
    
end


end

File: cpp/src/dijkstra_4dir.cpp
#include "dijkstra_4dir.h"
#include "map.h"
#include <vector>
#include <utility>
#include <queue>
#include <iostream>
#include <math.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

std::vector<std::pair<double, double>> dijkstra_4dir( Map* theMap, 
	const std::pair<double, double>& start_loc, const std::pair<double, double>& goal_loc) {

	std::vector<std::pair<double, double> > initial_path;
	std::queue<std::pair<int, int>> pos_queue;
	pos_queue.push(std::pair<int, int>(floor(start_loc.first), floor(start_loc.second)));

	int* Mmap = theMap->data_;
	int xsize = theMap->xsize_;

	std::vector<int> father_index;
	father_index.resize(theMap->xsize_*theMap->ysize_, -1);

	std::vector<double> cost;
	cost.resize(theMap->xsize_*theMap->ysize_, 99999.0);
	cost[floor(start_loc.second)*xsize + floor(start_loc.first)] = 0;

	std::vector<std::pair<int, int>> sons;
	std::vector<double> theta_vec;

	while (1) {
		if (pos_queue.empty()) {
			std::cout << "dijkstra failed" << std::endl;
			break;
		}

		int x = pos_queue.front().first;
		int y = pos_queue.front().second;

		pos_queue.pop();

		sons.clear();
		theta_vec.clear();
		std::pair<int, int> son_pos;

		son_pos.first = x - 1;
		son_pos.second = y;
		sons.emplace_back(son_pos);
		theta_vec.emplace_back(M_PI);

		son_pos.first = x;
		son_pos.second = y - 1;
		sons.emplace_back(son_pos);
		theta_vec.emplace_back(-M_PI/2);

		son_pos.first = x;
		son_pos.second = y + 1;
		sons.emplace_back(son_pos);
		theta_vec.emplace_back(M_PI/2);

		son_pos.first = x + 1;
		son_pos.second = y;
		sons.emplace_back(son_pos);
		theta_vec.emplace_back(0);

		if (x == floor(goal_loc.first) && y == floor(goal_loc.second)) {
			std::cout << "Dijkstra: find path for initial tether state estimation" << std::endl;

			initial_path.clear();

			std::pair<int, int> cur;
			cur.first = x;
			cur.second = y;
			while (true) {
				if (cur.first == floor(start_loc.first) && cur.second == floor(start_loc.second))
				{
					initial_path.insert(initial_path.begin(), std::pair<double, double>(cur.first + 0.5, cur.second + 0.5));
					break;
				}
				int cur_index = cur.second * xsize + cur.first;
				int fatherindex = father_index[cur_index];
				int father_x = fatherindex % xsize;
				int father_y = (fatherindex - father_x) / xsize;

				initial_path.insert(initial_path.begin(), std::pair<double, double>(cur.first + 0.5, cur.second + 0.5));
				cur.first = father_x;
				cur.second = father_y;
			}
			break;
		}

		for (int i = 0; i < sons.size(); i++) {
			std::pair<int, int> son_pos = sons[i];

			if (!theMap->isCollisionFree(son_pos.first, son_pos.second, theta_vec[i])) {
				continue;
			}

			if (cost[son_pos.second*xsize + son_pos.first] >= 0 && 
				cost[son_pos.second*xsize + son_pos.first] < 9998) 
			{
				// We are using BFS
				// If a node has been explored, then we cannot have a shortcut
				continue;
			}

			int son_index = son_pos.second *xsize + son_pos.first;
			int fatherindex = y * xsize + x;
			father_index[son_index] = fatherindex;
			cost[son_index] = cost[fatherindex] + 1;

			pos_queue.push(son_pos);
		}
	}
	return initial_path;
}